File: backend/src/services/spotify.js
const axios = require('axios');
const NodeCache = require('node-cache');
const dotenv = require('dotenv');
const path = require('path');

// Load environment variables
dotenv.config({ path: path.join(__dirname, '../../.env') });

const cache = new NodeCache({ stdTTL: 3600 }); // 1 hour cache
const TOKEN_CACHE_KEY = 'spotify_access_token';

class SpotifyService {
  constructor() {
    this.clientId = process.env.SPOTIFY_CLIENT_ID;
    this.clientSecret = process.env.SPOTIFY_CLIENT_SECRET;
    this.tokenUrl = 'https://accounts.spotify.com/api/token';
    this.baseUrl = 'https://api.spotify.com/v1';
  }

  async getAccessToken() {
    const cachedToken = cache.get(TOKEN_CACHE_KEY);
    if (cachedToken) {
      return cachedToken;
    }

    try {
      const response = await axios.post(
        this.tokenUrl,
        'grant_type=client_credentials',
        {
          headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
            'Authorization': 'Basic ' + Buffer.from(
              this.clientId + ':' + this.clientSecret
            ).toString('base64')
          }
        }
      );

      const token = response.data.access_token;
      // Cache token for 50 minutes (expires in 1 hour)
      cache.set(TOKEN_CACHE_KEY, token, 3000);
      return token;
    } catch (error) {
      console.error('Spotify token error:', error.message);
      throw error;
    }
  }

  async searchArtist(artistName) {
    const cacheKey = `artist_${artistName}`;
    const cachedData = cache.get(cacheKey);
    
    if (cachedData) {
      return cachedData;
    }

    try {
      const token = await this.getAccessToken();
      const response = await axios.get(`${this.baseUrl}/search`, {
        params: {
          q: artistName,
          type: 'artist',
          limit: 1
        },
        headers: {
          'Authorization': `Bearer ${token}`
        }
      });

      const artist = response.data.artists.items[0];
      cache.set(cacheKey, artist);
      return artist;
    } catch (error) {
      console.error(`Spotify artist search error (${artistName}):`, error.message);
      return null;
    }
  }

  async searchAlbum(albumName, artistName) {
    const cacheKey = `album_${albumName}_${artistName}`;
    const cachedData = cache.get(cacheKey);
    
    if (cachedData) {
      return cachedData;
    }

    try {
      const token = await this.getAccessToken();
      const response = await axios.get(`${this.baseUrl}/search`, {
        params: {
          q: `album:${albumName} artist:${artistName}`,
          type: 'album',
          limit: 1
        },
        headers: {
          'Authorization': `Bearer ${token}`
        }
      });

      const album = response.data.albums.items[0];
      cache.set(cacheKey, album);
      return album;
    } catch (error) {
      console.error(`Spotify album search error (${albumName}):`, error.message);
      return null;
    }
  }

  async searchTrack(trackName, artistName) {
    const cacheKey = `track_${trackName}_${artistName}`;
    const cachedData = cache.get(cacheKey);
    
    if (cachedData) {
      return cachedData;
    }

    try {
      const token = await this.getAccessToken();
      const response = await axios.get(`${this.baseUrl}/search`, {
        params: {
          q: `track:${trackName} artist:${artistName}`,
          type: 'track',
          limit: 1
        },
        headers: {
          'Authorization': `Bearer ${token}`
        }
      });

      const track = response.data.tracks.items[0];
      cache.set(cacheKey, track);
      return track;
    } catch (error) {
      console.error(`Spotify track search error (${trackName}):`, error.message);
      return null;
    }
  }
}

module.exports = new SpotifyService();

File: backend/src/services/spotify.rs
use std::sync::Arc;
use std::time::Duration;

use anyhow::{Context, Result};
use base64::{engine::general_purpose::STANDARD as BASE64, Engine as _};
use moka::future::Cache;
use reqwest::Client;
use serde::Deserialize;
use serde_json::Value;
use tokio::sync::RwLock;
use tracing::debug;

use crate::config::Config;

const SPOTIFY_TOKEN_URL: &str = "https://accounts.spotify.com/api/token";
const SPOTIFY_BASE_URL: &str = "https://api.spotify.com/v1";

#[derive(Deserialize)]
struct TokenResponse {
    access_token: String,
}

#[derive(Clone)]
pub struct SpotifyService {
    client: Client,
    search_cache: Cache<String, Value>,
    /// Cached OAuth token + its expiry instant
    token: Arc<RwLock<Option<(String, std::time::Instant)>>>,
    client_id: String,
    client_secret: String,
    token_ttl: Duration,
}

impl SpotifyService {
    pub fn new(config: &Config) -> Self {
        let client = Client::builder()
            .timeout(Duration::from_secs(10))
            .pool_max_idle_per_host(10)
            .build()
            .expect("failed to build HTTP client");

        let search_cache = Cache::builder()
            .max_capacity(5_000)
            .time_to_live(Duration::from_secs(config.spotify_cache_ttl_secs))
            .build();

        Self {
            client,
            search_cache,
            token: Arc::new(RwLock::new(None)),
            client_id: config.spotify_client_id.clone(),
            client_secret: config.spotify_client_secret.clone(),
            token_ttl: Duration::from_secs(config.spotify_token_ttl_secs),
        }
    }

    /// Returns a valid access token, refreshing when necessary.
    async fn access_token(&self) -> Result<String> {
        // Fast path — read lock only
        {
            let guard = self.token.read().await;
            if let Some((token, expires_at)) = guard.as_ref() {
                if std::time::Instant::now() < *expires_at {
                    return Ok(token.clone());
                }
            }
        }

        // Slow path — acquire write lock and refresh
        let mut guard = self.token.write().await;
        // Double-check after acquiring write lock
        if let Some((token, expires_at)) = guard.as_ref() {
            if std::time::Instant::now() < *expires_at {
                return Ok(token.clone());
            }
        }

        debug!("refreshing Spotify access token");
        let credentials = BASE64.encode(format!("{}:{}", self.client_id, self.client_secret));
        let resp = self
            .client
            .post(SPOTIFY_TOKEN_URL)
            .header("Authorization", format!("Basic {credentials}"))
            .header("Content-Type", "application/x-www-form-urlencoded")
            .body("grant_type=client_credentials")
            .send()
            .await
            .context("Spotify token request failed")?
            .error_for_status()
            .context("Spotify token endpoint returned error")?
            .json::<TokenResponse>()
            .await
            .context("failed to parse Spotify token response")?;

        let expires_at = std::time::Instant::now() + self.token_ttl;
        *guard = Some((resp.access_token.clone(), expires_at));
        Ok(resp.access_token)
    }

    async fn search(&self, query: &str, search_type: &str) -> Result<Option<Value>> {
        let cache_key = format!("{search_type}::{query}");
        if let Some(cached) = self.search_cache.get(&cache_key).await {
            debug!("cache hit: {cache_key}");
            return Ok(Some(cached));
        }

        let token = self.access_token().await?;
        let mut resp: Value = self
            .client
            .get(format!("{SPOTIFY_BASE_URL}/search"))
            .bearer_auth(&token)
            .query(&[("q", query), ("type", search_type), ("limit", "1")])
            .send()
            .await
            .context("Spotify search request failed")?
            .error_for_status()
            .context("Spotify search returned error status")?
            .json()
            .await
            .context("failed to parse Spotify search JSON")?;

        // Pull the first item from the appropriate list
        let key = format!("{}s", search_type); // e.g. "artists", "albums", "tracks"
        let item = resp[key]["items"]
            .as_array_mut()
            .and_then(|arr| if arr.is_empty() { None } else { Some(arr.remove(0)) });

        if let Some(ref v) = item {
            self.search_cache.insert(cache_key, v.clone()).await;
        }
        Ok(item)
    }

    pub async fn search_artist(&self, name: &str) -> Result<Option<Value>> {
        self.search(name, "artist").await
    }

    pub async fn search_album(&self, album: &str, artist: &str) -> Result<Option<Value>> {
        let q = format!("album:{album} artist:{artist}");
        self.search(&q, "album").await
    }

    pub async fn search_track(&self, track: &str, artist: &str) -> Result<Option<Value>> {
        let q = format!("track:{track} artist:{artist}");
        self.search(&q, "track").await
    }
}

pub type SharedSpotify = Arc<SpotifyService>;
